### PR TITLE
Add platform RDK to launcher framework

### PR DESCRIPTION
**Pull request by maxz-lab**
Added launcher infrastructure for AH212 RDK (v5) as independent RDK platform.

b/293172465

### CODE REVIEW

File: starboard/rdk/launcher.py [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+
+    # Ensure no trailing slashes
+    self.out_directory = self.out_directory.rstrip('/')
+
+    rdk_user_hostname = Launcher._RDK_USERNAME + '@' + self.device_id
+    rdk_test_dir = '/usr/share/content/data/app'
+
+    # scp command setup
+    options = '-rOCq '
+    rsa_options = '-o \"LogLevel ERROR\" \
```
Comment (oxve) [inline starboard/rdk/launcher.py:143]: Don't use backslash for line continuation: go/pystyle#line-length

There's some more of these below.

File: starboard/rdk/launcher.py [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+  def _InitPexpectCommands(self):
+    """Initializes all of the pexpect commands needed for running the test."""
+
+    # Ensure no trailing slashes
+    self.out_directory = self.out_directory.rstrip('/')
+
+    rdk_user_hostname = Launcher._RDK_USERNAME + '@' + self.device_id
+    rdk_test_dir = '/usr/share/content/data/app'
+
+    # scp command setup
```
Comment (oxve) [inline starboard/rdk/launcher.py:141]: Can you use a format strings for all of the strings below?

File: starboard/rdk/launcher.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+    self.test_failure_list_tag = 'FAILED TESTS'
+
+    # test command setup
+    # note: If RDK has ResidentApp feature turned on (default), the log
+    #       goes to a buffer that can be retrieved by command
+    #           journalctl -f -u dobby
+    #       Otherwise, the container is off, and the log is dumped into
+    #       file /opt/logs/wpeframework.log.
+    #           tail -f /opt/logs/wpeframework.log
+    self.test_command = 'curl -X POST http://127.0.0.1:9998/jsonrpc -d \
```
Comment (oxve) [inline starboard/rdk/launcher.py:167]: Can you break out the values in this string to variables or constants (as applicable)?

File: starboard/tools/testing/test_runner.py [hunk 4]
```diff
@@ -40,11 +40,11 @@
 # pylint: disable=consider-using-f-string
 
 _FLAKY_RETRY_LIMIT = 4
-_TOTAL_TESTS_REGEX = re.compile(r"^\[==========\] (.*) tests? from .*"
```
Comment (oxve) [inline starboard/tools/testing/test_runner.py:43]: Why are you removing these?

File: starboard/tools/testing/test_runner.py [hunk 5]
```diff
@@ -265,7 +265,7 @@ def __init__(self,
     _VerifyConfig(self._platform_config,
                   self._platform_test_filters.GetTestFilters())
 
-    if self.loader_platform:
+    if self.loader_platform and self.loader_platform.find("rdk") == -1:
```
Comment (oxve) [inline starboard/tools/testing/test_runner.py:268]: We don't want platform specific code like this in the test runner. I think the best way to handle this is probably to add some logic to the build config (`gyp_configuration.py`). It's fetched about on line 250.

File: starboard/tools/testing/test_runner.py [hunk 6]
```diff
@@ -786,7 +786,7 @@ def BuildAllTargets(self, ninja_flags):
 
       # The loader is not built with the same platform configuration as our
       # tests so we need to build it separately.
-      if self.loader_platform:
+      if self.loader_platform and self.loader_platform.find("rdk") == -1:
```
Comment (oxve) [inline starboard/tools/testing/test_runner.py:789]: Same comment about platform specific code here

File: starboard/tools/testing/test_runner.py (hunk 4 above)
Comment (maxz-lab) [reply starboard/tools/testing/test_runner.py:43]: The output from RDK has a prefix string in front of our output. 
For example:
Sep 11 23:23:36.071628 WPEFramework[20227]: [  PASSED  ] 2664 tests.
Sep 11 23:23:36.071628 WPEFramework[20227]: [  FAILED  ] 75 tests, listed

File: starboard/evergreen/shared/launcher.py [hunk 7]
```diff
@@ -175,6 +175,8 @@ def _StageTargetsAndContents(self):
     # TODO(b/267568637): Make the Linux launcher run from the install_directory.
     if 'linux' in self.loader_platform:
       self._StageTargetsAndContentsLinux()
+    elif 'rdk' in self.loader_platform:
+      self._StageTargetsAndContentsRdk()
     else:
```
[the remark this answers is not in the document]
Comment (maxz-lab) [reply starboard/evergreen/shared/launcher.py:180]: Yes. The check is added as a safety net. Before it throws error at this point, build.py (line 102) will report failed to load module error.

File: starboard/evergreen/shared/launcher.py [hunk 8]
```diff
@@ -271,6 +273,41 @@ def _StageTargetsAndContentsRaspi(self):
     os.makedirs(os.path.join(target_staging_dir, 'lib'))
     shutil.copy(target_binary_src, target_binary_dst)
 
+  def _StageTargetsAndContentsRdk(self):
+    """Stage targets and their contents for GN builds for RDK platforms."""
+    content_subdir = os.path.join('usr', 'share', 'cobalt')
+    #target_name should be self.target_name. However because loader is hardcoded
```
[the remark this answers is not in the document]
Comment (maxz-lab) [reply starboard/evergreen/shared/launcher.py:279]: corrected.

File: starboard/evergreen/shared/launcher.py [hunk 9]
```diff
@@ -271,6 +273,41 @@ def _StageTargetsAndContentsRaspi(self):
...
+    #to call "libcobalt.so" in path "cobalt", it has to use this name for now.
+    #b/296633713
```
[the remark this answers is not in the document]
Comment (maxz-lab) [reply starboard/evergreen/shared/launcher.py:281]: Yes. TODO keyword is added.

File: starboard/evergreen/shared/launcher.py [hunk 10]
```diff
@@ -271,6 +273,41 @@ def _StageTargetsAndContentsRaspi(self):
...
+    target_name = 'cobalt'
+
+    # Copy target content and binary
```
[the remark this answers is not in the document]
Comment (maxz-lab) [reply starboard/evergreen/shared/launcher.py:284]: Yes.

File: starboard/evergreen/shared/launcher.py [hunk 11]
```diff
@@ -271,6 +273,41 @@ def _StageTargetsAndContentsRaspi(self):
...
+    target_install_path = os.path.join(self.out_directory, 'install')
+    target_staging_dir = os.path.join(self.staging_directory, 'content', 'app',
+                                      target_name)
+    os.makedirs(target_staging_dir)
+
+    # TODO(b/218889313): Reset the content path for the evergreen artifacts.
```
[the remark this answers is not in the document]
Comment (maxz-lab) [reply starboard/evergreen/shared/launcher.py:290]: It is needed, because it is target independent.

File: starboard/rdk/launcher.py (hunk 2 above)
Comment (maxz-lab) [reply starboard/rdk/launcher.py:141]: Done.

File: starboard/tools/testing/test_runner.py (hunk 5 above)
Comment (maxz-lab) [reply starboard/tools/testing/test_runner.py:268]: Removed the platform specific check from test_runner.

File: starboard/tools/testing/test_runner.py (hunk 6 above)
Comment (maxz-lab) [reply starboard/tools/testing/test_runner.py:789]: Removed the platform specific check from test_runner.

**Comment by maxz-lab** [comment]
Updated a new patch to address all the comments from Oxve.

File: starboard/evergreen/shared/launcher.py [hunk 12]
```diff
@@ -271,6 +275,43 @@ def _StageTargetsAndContentsRaspi(self):
...
+    content_subdir = os.path.join('usr', 'share', 'cobalt')
+    target_content_src = os.path.join(target_install_path, content_subdir)
+    target_content_dst = os.path.join(target_staging_dir, 'content')
+    shutil.copytree(target_content_src, target_content_dst)
+
+    shlib_name = f'lib{self.target_name}'
+    shlib_name += '.lz4' if self.use_compressed_library else '.so'
```
Comment (oxve) [inline starboard/evergreen/shared/launcher.py:301]: This snippet is irrelevant for RDK right? The loader is hardcoded to load `libcobalt.so` so the `lz4` extension wouldn't work.

File: starboard/rdk/launcher.py (hunk 3 above)
Comment (oxve) [reply starboard/rdk/launcher.py:167]: I'm not asking you to change the content of the string, just make it more readable. The line currently has two separate commands that are lumped into one, the curl command (which I'm assuming is triggering the loading of `libcobalt.so` and the tail command to view the logs. It also contains some hardcoded (I assume?) json fields. 

These could be broken up into variables and put back together in a format string.

Comment (maxz-lab) [reply starboard/rdk/launcher.py:167]: Correct. I am working on that.

File: starboard/evergreen/shared/launcher.py (hunk 12 above)
Comment (maxz-lab) [reply starboard/evergreen/shared/launcher.py]: Removed the irrelevant 'lz4'.

File: starboard/rdk/launcher.py (hunk 3 above)
Comment (maxz-lab) [reply starboard/rdk/launcher.py]: Sorry that I didn't catch your idea yesterday. Here it is. Also once we sort out the argument issue with RDK, more variables will be added into the command line in this fashion.

File: starboard/rdk/retry.py
[the remark this answers is not in the document]
Comment (maxz-lab) [reply starboard/rdk/retry.py:1]: It is kind needed still. I have the retry module moved to shared folder and shared between RDK and raspi-2.

File: starboard/evergreen/shared/launcher.py [hunk 13]
```diff
@@ -271,6 +275,43 @@ def _StageTargetsAndContentsRaspi(self):
...
 
+  def _StageTargetsAndContentsRdk(self):
+    """Stage targets and their contents for GN builds for RDK platforms."""
+    content_subdir = os.path.join('usr', 'share', 'cobalt')
+
+    # TODO(b/296633713):
+    # Target_name should be self.target_name. However because loader is
+    # hardcoded on RDK, i.e. it always loads "libcobalt.so" in path "cobalt",
+    # it has to use this name.
+    target_name = 'cobalt'
```
Comment (andrewsavage1) [inline starboard/evergreen/shared/launcher.py:286]: Is there a reason to add this instead of just using `-t cobalt` and having other `-t` values fail?

File: starboard/rdk/launcher.py [hunk 14]
```diff
@@ -0,0 +1,426 @@
...
+import os
+import re
+import signal
+import six
+import sys
+import threading
+import time
+import contextlib
+
+import pexpect
```
Comment (andrewsavage1) [inline starboard/rdk/launcher.py:28]: Is pexpect something we have to use? @kaidokert I know you were looking at some alternative solutions, it seems like those could be more easily implemented on a new platform like this?

Comment (andrewsavage1) [reply starboard/rdk/launcher.py:28]: Looks like we have an unfulfilled MH depenency

File: starboard/evergreen/shared/launcher.py (hunk 13 above)
Comment (andrewsavage1) [reply starboard/evergreen/shared/launcher.py:286]: Looks like I misunderstood, once comment is updated lgtm

**Review by andrewsavage1 — APPROVED** [review]

File: starboard/evergreen/shared/launcher.py [hunk 15]
```diff
@@ -271,6 +275,43 @@ def _StageTargetsAndContentsRaspi(self):
     os.makedirs(os.path.join(target_staging_dir, 'lib'))
     shutil.copy(target_binary_src, target_binary_dst)
 
+  def _StageTargetsAndContentsRdk(self):
+    """Stage targets and their contents for GN builds for RDK platforms."""
+    content_subdir = os.path.join('usr', 'share', 'cobalt')
+
+    # TODO(b/296633713):
```
Comment (oxve) [inline starboard/evergreen/shared/launcher.py:282]: This bug is marked resolved.

I understand the context here a bit better now and it seems this is not a TODO. Can you confirm that there are no plans to fix the hardcoded file name in the RDK loader?

The comment could be a bit more concise, e.g.:

`The rdk loader always loads libcobalt.so. Hard code target_name to match.`

File: starboard/evergreen/shared/launcher.py [hunk 16]
```diff
@@ -271,6 +275,43 @@ def _StageTargetsAndContentsRaspi(self):
...
+    # Target_name should be self.target_name. However because loader is
+    # hardcoded on RDK, i.e. it always loads "libcobalt.so" in path "cobalt",
+    # it has to use this name.
+    target_name = 'cobalt'
+
+    # Copy target content and binary.
+    target_install_path = os.path.join(self.out_directory, 'install')
+    target_staging_dir = os.path.join(self.staging_directory, 'content', 'app',
+                                      target_name)
```
Comment (oxve) [inline starboard/evergreen/shared/launcher.py:291]: Does this actually work? The content path I think would not change due to the hard coded path in the loader.

Should the hardcoding of `target_name` happen below this line instead?

File: starboard/evergreen/shared/launcher.py [hunk 17]
```diff
@@ -271,6 +275,43 @@ def _StageTargetsAndContentsRaspi(self):
...
+    os.makedirs(target_staging_dir)
+
+    # TODO(b/218889313): Reset the content path for the evergreen artifacts.
+    content_subdir = os.path.join('usr', 'share', 'cobalt')
+    target_content_src = os.path.join(target_install_path, content_subdir)
+    target_content_dst = os.path.join(target_staging_dir, 'content')
+    shutil.copytree(target_content_src, target_content_dst)
+
+    shlib_name = f'lib{self.target_name}'
+    shlib_name += '.so'
```
Comment (oxve) [inline starboard/evergreen/shared/launcher.py:301]: Inline this dangling append above.

File: starboard/evergreen/shared/launcher.py [hunk 18]
```diff
@@ -271,6 +273,41 @@ def _StageTargetsAndContentsRaspi(self):
...
+    content_subdir = os.path.join('usr', 'share', 'cobalt')
+    target_content_src = os.path.join(target_install_path, content_subdir)
+    target_content_dst = os.path.join(target_staging_dir, 'content')
+    shutil.copytree(target_content_src, target_content_dst)
+
+    shlib_name = f'lib{self.target_name}'
+    shlib_name += '.lz4' if self.use_compressed_library else '.so'
+
+    target_binary_src = os.path.join(self.out_directory, shlib_name)
+    if target_name in shlib_name:
```
[the remark this answers is not in the document]
Comment (oxve) [reply starboard/evergreen/shared/launcher.py:304]: Right, so the else here is never entered. If there's no plan to change the RDK launcher to load other file names this else branch should be removed.

**Review by oxve — COMMENTED** [review]
The title of the PR doesn't really match the contents. Can you rephrase to make it clear that it adds a new platform to the launcher framework?

File: starboard/rdk/launcher.py [hunk 19]
```diff
@@ -0,0 +1,426 @@
...
+  def _KillExistingCobaltProcesses(self):
+    """If there are leftover Cobalt processes, kill them.
+
+    It is possible that a previous process did not exit cleanly.
+    Zombie Cobalt instances can block the WebDriver port or
+    cause other problems.
+    """
+    logging.info('Killing existing processes')
+    self._PexpectSendLine(
+        'pkill -9 -ef "(cobalt)|(crashpad_handler)|(elf_loader)"')
```
Comment (oxve) [inline starboard/rdk/launcher.py:313]: Does pkill work on RDK? In the 'Run local build Evergreen on RDK' doc you mention a terminate command. Should that be used instead?

File: starboard/rdk/launcher.py [hunk 20]
```diff
@@ -0,0 +1,426 @@
...
+
+    # scp command setup
+    options = '-rOCq'
+    rsa_options = (
+        '-o \"LogLevel ERROR\" '
+        '-o \"UserKnownHostsFile=/dev/null\" -o \"StrictHostKeyChecking=no\"')
+
+    source = os.path.join(self.out_directory, 'content', 'app', 'cobalt')
+    destination = f'{rdk_user_hostname}:{rdk_test_dir}/'
+    self.scp_command = (f'scp {options} {rsa_options} '
```
Comment (oxve) [inline starboard/rdk/launcher.py:149]: Why are you using scp instead of rsync to copy over the files? Is rsync not supported on the RDK device?

File: starboard/rdk/gyp_configuration.py [hunk 21]
```diff
@@ -0,0 +1,34 @@
+# Copyright 2017 The Cobalt Authors. All Rights Reserved.
```
Comment (oxve) [inline starboard/rdk/gyp_configuration.py:1]: Can you update the copyright year to 2023? For a new file it should have the current year.

File: starboard/rdk/launcher.py [hunk 22]
```diff
@@ -0,0 +1,426 @@
+#
+# Copyright 2018 The Cobalt Authors. All Rights Reserved.
```
Comment (oxve) [inline starboard/rdk/launcher.py:2]: Update year here too

File: starboard/rdk/test_filters.py [hunk 23]
```diff
@@ -0,0 +1,35 @@
+# Copyright 2022 The Cobalt Authors. All Rights Reserved.
```
Comment (oxve) [inline starboard/rdk/test_filters.py:1]: Another year.

File: starboard/rdk/launcher.py (hunk 20 above)
Comment (maxz-lab) [reply starboard/rdk/launcher.py:149]: Switched back to rsync.

File: starboard/evergreen/shared/launcher.py (hunk 16 above)
Comment (maxz-lab) [reply starboard/evergreen/shared/launcher.py:291]: This is to form the local staging folder, which will be mirrored/rsync to RDK. The "cobalt" is needed because the path is hardcoded as well.

File: starboard/evergreen/shared/launcher.py [hunk 24]
```diff
@@ -271,6 +275,33 @@ def _StageTargetsAndContentsRaspi(self):
...
+    os.makedirs(target_staging_dir)
+
+    # TODO(b/218889313): Reset the content path for the evergreen artifacts.
+    content_subdir = os.path.join('usr', 'share', 'cobalt')
+    target_content_src = os.path.join(target_install_path, content_subdir)
+    target_content_dst = os.path.join(target_staging_dir, 'content')
+    shutil.copytree(target_content_src, target_content_dst)
+
+    shlib_name = f'lib{self.target_name}.so'
+
```
Comment (oxve) [inline starboard/evergreen/shared/launcher.py:298]: nit: remove this whitespace

File: starboard/rdk/test_filters.py [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Starboard RDK Platform Test Filters."""
+
+from starboard.tools.testing import test_filter
+
+# pylint: disable=line-too-long
+_FILTERED_TESTS = {
```
Comment (oxve) [inline starboard/rdk/test_filters.py:19]: Can you add TODOs for any bugs that are tracking the work to fix this?

**Review by oxve — APPROVED** [review]

File: starboard/rdk/test_filters.py (hunk 25 above)
Comment (maxz-lab) [reply starboard/rdk/test_filters.py:19]: b/302394149
